<a href="https://colab.research.google.com/github/hsarfraz/Tiny-Machine-Learning/blob/main/0_8_DeepNN_vs_ConvolutionNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Improving Computer Vision Accuracy using Convolutions

In previous lessons you saw how to do image recognition using a Deep Neural Network (DNN) containing three layers:

1.  the input layer (in the shape of the data)
2.  the output layer (in the shape of the desired output)
3.  and a hidden layer


For convenience, here's the entire code again. Run it and take a note of the accuracy that is printed out at the end.

In [ ]:
import tensorflow as tf
import time

start = time.time()
mnist = tf.keras.datasets.fashion_mnist
(training_images, training_labels), (val_images, val_labels) = mnist.load_data() #creating mnist training, validation images

#feature scaling (normalising on training and validation images) (notebook 0.6)
training_images=training_images / 255.0
val_images=val_images / 255.0

#DNN -deep neural network
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(), #input layer that flattens image to 'flattened image #' by 1 (notebook 0.6)
  tf.keras.layers.Dense(20, activation=tf.nn.relu), #hidden layer w/ dense connection (notebook 0.3)
  tf.keras.layers.Dense(10, activation=tf.nn.softmax) #output layer
])

#model compiling is performed after defining the neural net model and before model training starts.
model.compile(optimizer='adam', #Optimizers are used to update weights and biases (model internal parameters) to reduce the error/loss
              loss='sparse_categorical_crossentropy', #loss function to calculate loss
              metrics=['accuracy'] #A metric is a function that is used to judge the performance of your model
              )

#Model Fitting is the process of training the neural network to figure out the relationship between X and Y (notebook 0.5)
model.fit(training_images, training_labels, validation_data=(val_images, val_labels), epochs=20
          ,verbose=0
          )

end = time.time()
time_in_seconds = end - start
print('Number of minutes it took train the DNN: ', time_in_seconds/60)

Number of minutes it took train the DNN:  2.392892054716746


Here is the screenshot of the DNN training results (I used `verbose=0` in the `model.fit()` function to hide the epoch outputs so it does not take up too much notebook space:

<img src="images/0.8_DNN_training_results.jpg" width="700">

Your accuracy is probably about 89% on training and 87% on validation...not bad...But how do you make that even better? One way is to use something called Convolutions. I'm not going to get into the details of Convolutions here, but the ultimate concept is that they narrow down the content of the image to focus on specific, distinct, features.

If you've ever done image processing using a filter (like this: https://en.wikipedia.org/wiki/Kernel_(image_processing)) then convolutions will look very familiar.

In short, you take an array (usually 3x3 or 5x5) and pass it over the image. By changing the underlying pixels based on the formula within that matrix, you can do things like edge detection. So, for example, if you look at the above link, you'll see a 3x3 that is defined for edge detection where the middle cell is 8, and all of its neighbors are -1. In this case, for each pixel, you would multiply its value by 8, then subtract the value of each neighbor. Do this for every pixel, and you'll end up with a new image that has the edges enhanced.

This is perfect for computer vision, because often it's features like edges that distinguish one item for another. And once we move from raw image data to feature data, the amount of information needed is then much less...because you'll just train on the highlighted features.

That's the concept of Convolutional Neural Networks. Add some layers to do convolution before you have the dense layers, and then the information going to the dense layers is more focussed, and possibly more accurate.

Run the below code -- this is the same neural network as earlier, but this time with Convolutional layers added first. It will take longer, but look at the impact on the accuracy:

In [ ]:
import tensorflow as tf
import time

start = time.time()
mnist = tf.keras.datasets.fashion_mnist
(training_images, training_labels), (val_images, val_labels) = mnist.load_data()

#feature scaling (normalising on training and validation images) (notebook 0.6)
training_images=training_images.reshape(60000, 28, 28, 1) #making the image items into one big 4D list item
training_images=training_images / 255.0
val_images=val_images.reshape(10000, 28, 28, 1)
val_images=val_images/255.0

#CNN -convolution neural network
model = tf.keras.models.Sequential([
  #convolution number 1
  tf.keras.layers.Conv2D(64, #number of convolutions
                         (3,3), #size of convolution
                         activation='relu', #activation function
                         input_shape=(28, 28, 1) #the shape of the input data in the first layer
                         ),

  #pooling compresses the image to highlight the image features (notebook 0.8)
  tf.keras.layers.MaxPooling2D(2, 2) #the pooling effect is to reduce the image size to a quarter
  ,
  tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
  tf.keras.layers.MaxPooling2D(2,2),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(20, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax')
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.summary()
model.fit(training_images, training_labels, validation_data=(val_images, val_labels), epochs=20, verbose=0)

end = time.time()
time_in_seconds = end - start
print('Number of minutes it took train the DNN: ', time_in_seconds/60)

4422102/4422102 [==============================] - 0s 0us/step
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 64)        640       
                                                                 
 max_pooling2d (MaxPooling2  (None, 13, 13, 64)        0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        36928     
                                                                 
 max_pooling2d_1 (MaxPoolin  (None, 5, 5, 64)          0         
 g2D)                                                            
                                                                 
 flatten (Flatten)           (None, 1600)              0         
                                                           

Here is a screenshot of the CNN training results:

<img src="images/0.8_CNN_Training_Results.jpg" width="700">

It's likely gone up to about 97% on the training data and 91% on the validation data.

That's significant, and a step in the right direction!

Try running it for more epochs -- say about 100, and explore the results! But while the results might seem really good, the validation results may actually go down, due to something called 'overfitting' which will be discussed later.

(In a nutshell, 'overfitting' occurs when the network learns the data from the training set really well, but it's too specialised to only that data, and as a result is less effective at seeing *other* data. For example, if all your life you only saw red shoes, then when you see a red shoe you would be very good at identifying it, but blue suade shoes might confuse you...and you know you should never mess with my blue suede shoes.)

Then, look at the code again, and see, step by step how the Convolutions were built:

Step 1 is to gather the data. You'll notice that there's a bit of a change here in that the training data needed to be reshaped. That's because the first convolution expects a single tensor containing everything, so instead of 60,000 28x28x1 items in a list, we have a single 4D list that is 60,000x28x28x1, and the same for the validation images. If you don't do this, you'll get an error when training as the Convolutions do not recognize the shape.



```
import tensorflow as tf
mnist = tf.keras.datasets.fashion_mnist
(training_images, training_labels), (val_images, val_labels) = mnist.load_data()
training_images=training_images.reshape(60000, 28, 28, 1)
training_images=training_images / 255.0
val_images = val_images.reshape(10000, 28, 28, 1)
val_images=val_images/255.0
```



Next is to define your model. Now instead of the input layer at the top, you're going to add a Convolution. The parameters are:

1. The number of convolutions you want to generate. Purely arbitrary, but good to start with something in the order of 64
2. The size of the Convolution, in this case a 3x3 grid
3. The activation function to use -- in this case we'll use relu, which you might recall is the equivalent of returning x when x>0, else returning 0
4. In the first layer, the shape of the input data.

You'll follow the Convolution with a MaxPooling layer which is then designed to compress the image, while maintaining the content of the features that were highlighted by the convolution. By specifying (2,2) for the MaxPooling, the effect is to quarter the size of the image. Without going into too much detail here, the idea is that it creates a 2x2 array of pixels, and picks the biggest one, thus turning 4 pixels into 1. It repeats this across the image, and in so doing halves the number of horizontal, and halves the number of vertical pixels, effectively reducing the image to 25% of its original size.

You can call model.summary() to see the size and shape of the network, and you'll notice that after every MaxPooling layer, the image size is reduced in this way.


```
model = tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(64, (3,3), activation='relu', input_shape=(28, 28, 1)),
  tf.keras.layers.MaxPooling2D(2, 2),
```



Add another convolution



```
  tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
  tf.keras.layers.MaxPooling2D(2,2)
```



Now flatten the output. After this you'll just have the same DNN structure as the non convolutional version

```
  tf.keras.layers.Flatten(),
```



The same 20 dense layers, and 10 output layers as in the pre-convolution example:



```
  tf.keras.layers.Dense(20, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax')
])
```



Now compile the model, call the fit method to do the training, and evaluate the loss and accuracy from the validation set.



```
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(training_images, training_labels, validation_data=(val_images, val_labels), epochs=20)
```




# Optional Additonal Exercises

## 1. Try editing the convolutions. Change the 32s to either 16 or 64. Explore what impact this will have on accuracy and/or training time.

After decreasing the number of convolutions the training and validation dataset accuracy decreased from 97% -> 92% and 91% -> 90% respectively. The training time decreased from 2.44 min -> 2.39 min.

<img src="images/0.8_CNN_16_convolutions.jpg" width="700">

In [ ]:
start = time.time()

#feature scaling (normalising on training and validation images) (notebook 0.6)
training_images=training_images.reshape(60000, 28, 28, 1) #making the image items into one big 4D list item
training_images=training_images / 255.0
val_images=val_images.reshape(10000, 28, 28, 1)
val_images=val_images/255.0

#CNN -convolution neural network
model = tf.keras.models.Sequential([
  #convolution number 1
  tf.keras.layers.Conv2D(16, #number of convolutions
                         (3,3), #size of convolution
                         activation='relu', #activation function
                         input_shape=(28, 28, 1) #the shape of the input data in the first layer
                         ),

  #pooling compresses the image to highlight the image features (notebook 0.8)
  tf.keras.layers.MaxPooling2D(2, 2) #the pooling effect is to reduce the image size to a quarter
  ,
  tf.keras.layers.Conv2D(16, (3,3), activation='relu'),
  tf.keras.layers.MaxPooling2D(2,2),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(20, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax')
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(training_images, training_labels, validation_data=(val_images, val_labels), epochs=20, verbose=0)

end = time.time()
time_in_seconds = end - start
print('Number of minutes it took train the DNN: ', time_in_seconds/60)

Number of minutes it took train the DNN:  2.3932920575141905


## 2. Remove the final Convolution. What impact will this have on accuracy or training time?

After removing the final convolution the training dataset accuracy increased from from 97% -> 98% and the validation dataset accuracy stayed the same 91% -> 91%. The training time decreased from 2.44 min -> 2.39 min.

<img src="images/0.8_removing_final_convolution.jpg" width="700">

In [ ]:
start = time.time()

#feature scaling (normalising on training and validation images) (notebook 0.6)
training_images=training_images.reshape(60000, 28, 28, 1) #making the image items into one big 4D list item
training_images=training_images / 255.0
val_images=val_images.reshape(10000, 28, 28, 1)
val_images=val_images/255.0

#CNN -convolution neural network
model = tf.keras.models.Sequential([
  #convolution number 1
  tf.keras.layers.Conv2D(64, #number of convolutions
                         (3,3), #size of convolution
                         activation='relu', #activation function
                         input_shape=(28, 28, 1) #the shape of the input data in the first layer
                         ),

  #pooling compresses the image to highlight the image features (notebook 0.8)
  tf.keras.layers.MaxPooling2D(2, 2) #the pooling effect is to reduce the image size to a quarter
  ,
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(20, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax')
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(training_images, training_labels, validation_data=(val_images, val_labels), epochs=20, verbose=0)

end = time.time()
time_in_seconds = end - start
print('Number of minutes it took train the DNN: ', time_in_seconds/60)

Number of minutes it took train the DNN:  2.386035672823588


## 3. How about adding more Convolutions? What impact do you think this will have? Experiment with it.

After adding a additional convolution layer the training dataset accuracy drastically decreased from from 97% -> 9% and the validation dataset accuracy stayed the same throughout every epoch and stayed at 10%. The training time decreased from 2.44 min -> 2.27 min.

<img src="images/0.8_additional_convolution_layer.jpg" width="700">

In [ ]:
start = time.time()

#feature scaling (normalising on training and validation images) (notebook 0.6)
training_images=training_images.reshape(60000, 28, 28, 1) #making the image items into one big 4D list item
training_images=training_images / 255.0
val_images=val_images.reshape(10000, 28, 28, 1)
val_images=val_images/255.0

#CNN -convolution neural network
model = tf.keras.models.Sequential([
  #convolution number 1
  tf.keras.layers.Conv2D(64, #number of convolutions
                         (3,3), #size of convolution
                         activation='relu', #activation function
                         input_shape=(28, 28, 1) #the shape of the input data in the first layer
                         ),

  #pooling compresses the image to highlight the image features (notebook 0.8)
  tf.keras.layers.MaxPooling2D(2, 2) #the pooling effect is to reduce the image size to a quarter
  ,
  tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
  tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
  tf.keras.layers.MaxPooling2D(2,2),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(20, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax')
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(training_images, training_labels, validation_data=(val_images, val_labels), epochs=20, verbose = 0)

end = time.time()
time_in_seconds = end - start
print('Number of minutes it took train the DNN: ', time_in_seconds/60)

Number of minutes it took train the DNN:  2.269674265384674
